datatypes/src/raster/grid_dimension.rs

```diff
@@ -123,301 +201,687 @@ impl<D> Capacity for D
+    }
+
+    fn intersection(&self, other: &Self) -> Self {
+        let [self_z, self_y, self_x] = self.size_as_index();
+        let [other_z, other_y, other_x] = other.size_as_index();
+        Dim3D::new([
+            min(self_z, other_z),
+            min(self_y, other_y),
+            min(self_x, other_x),
+        ])
+    }
+}
+
+impl From<[usize; 3]> for Dim3D {
+    #[inline]
+    fn from(dimension_size: [usize; 3]) -> Self {
+        Dim3D::new(dimension_size)
     }
 }
 
 impl Index<usize> for Dim3D {
     type Output = usize;
+    #[inline]
     fn index(&self, index: usize) -> &Self::Output {
         &self.dimension_size()[index]
     }
 }
 
 impl IndexMut<usize> for Dim3D {
+    #[inline]
     fn index_mut(&mut self, index: usize) -> &mut Self::Output {
         &mut self.dimension_size_mut()[index]
     }
 }
 
-impl Add for Dim3D {
-    type Output = Self;
+pub trait SignedGridIndex: Debug + Sized + AsRef<[SignedIdx]> {
+    type UnsignedIndex: GridIndex;
+
+    /// # Errors
+    /// This method fails if the grid index is out of bounds.
+    ///
+    fn signed_index_to_zero_based<D>(&self, dim: &D) -> Result<D::UnsignedIndexType>
+    where
+        D: OffsetDimension<SignedIndexType = Self, UnsignedIndexType = Self::UnsignedIndex>,
+    {
+        dim.offset_index_to_zero_based(self)
+    }
+
+    fn signed_index_to_zero_based_unchecked<D>(&self, dim: &D) -> D::UnsignedIndexType
+    where
+        D: OffsetDimension<SignedIndexType = Self, UnsignedIndexType = Self::UnsignedIndex>,
+    {
+        dim.offset_index_to_zero_based_unchecked(self)
+    }
+    fn inside_offset_dimension<D>(&self, dim: &D) -> bool
+    where
+        D: OffsetDimension<SignedIndexType = Self, UnsignedIndexType = Self::UnsignedIndex>,
+    {
+        dim.offset_index_inside_dimension(self)
+    }
 
-    fn add(self, rhs: Self) -> Self::Output {
-        let [z, y, x] = self.dimension_size;
-        let [other_z, other_y, other_x] = rhs.dimension_size;
-        Dim::new([z + other_z, y + other_y, x + other_x])
+    /// converts an offset index directly into a zero based linear space index
+    fn linear_space_index<D>(&self, dim: &D) -> Result<usize>
+    where
+        D: OffsetDimension<SignedIndexType = Self, UnsignedIndexType = Self::UnsignedIndex>,
+    {
+        dim.offset_index_to_linear_space_index(self)
     }
+
+    /// converts an offset index directly into a zero based linear space index
+    fn linear_space_index_unchecked<D>(&self, dim: &D) -> usize
+    where
+        D: OffsetDimension<SignedIndexType = Self, UnsignedIndexType = Self::UnsignedIndex>,
+    {
+        dim.offset_index_to_linear_space_index_unchecked(self)
+    }
+
+    fn add(&self, rhs: &Self) -> Self;
+    fn add_scalar(&self, rhs: SignedIdx) -> Self;
+    fn sub(&self, rhs: &Self) -> Self;
+    fn sub_scalar(&self, rhs: SignedIdx) -> Self;
+    fn mul(&self, rhs: &Self) -> Self;
+    fn mul_scalar(&self, rhs: SignedIdx) -> Self;
```

Comment:
Isn't it advantageous to use the default operator traits? This trait can then just expect that add, mul, etc. is implemented.

Reply:
would love to do this.

Reply:
There  is now a GridIdx type which can implement all the operations